Suggest mistyped CLI flag names based on ASCII or levenshtein distance

smileytechguy opened the pull request:
Fixes #1862

smileytechguy commented:
This will use, for > 1 character incorrect flags, levenshtein distance, showing the first one or two flags which match with a distance of <= 1 (you can see sample distances at [this site](https://planetcalc.com/1721/)).

For single character incorrect flags, the flag is compared against valid single-character flags by its ASCII character number.  It will show the first one or two flags which are less than three characters apart.

TysonAndre commented on src/Phan/CLI.php:
```diff
@@ -476,7 +476,105 @@ public function __construct()
+
+                    // if outside the for loop for efficiency
+                    if (strlen($key) == 1) {
+                        foreach ($knownFlags as $flag) {
```
I'm not sure what type of suggestion this is trying to make.

I also don't see much of a point in suggesting alternatives for single-character flags, and I'm not aware of other programs that do that.

TysonAndre replied:
An exception I just thought of was suggesting `-h` instead of `-H`, which was fixed on master (the comment was referring to avoiding suggestions such as  `--help` for `-P`

TysonAndre commented on src/Phan/CLI.php:
```diff
@@ -476,7 +476,105 @@ public function __construct()
                     break;
                 default:
-                    $this->usage("Unknown option '-$key'", EXIT_FAILURE);
+                    $knownFlags = [
```
This duplicates the above array, and they'll get out of sync

The flags array passed to `getopt` can be reused (by array_map(function(string $s) { return rtrim($s, ':'); })

- The flags array should probably be moved into an `@internal` class const

smileytechguy replied:
Sounds good

TysonAndre commented on src/Phan/CLI.php:
```diff
@@ -476,7 +476,105 @@ public function __construct()
+                        }
+                    }
+
+                    arsort($similarities); // retain keys and sort descending
```
This looks like it gives the user the **least** similar suggestions (e.g. if there were distances 5, 4, and 3, it would return the flags with distances 5 and 4, but not 3 (3 being the most suggestion)).

asort would sort as [0=>3,1=>4,2=>5], arsort as [0=>5,1=>4,2=>3]

smileytechguy replied:
Whoops, left the reverse sort on by accident from a previous revision.

TysonAndre commented on src/Phan/CLI.php:
```diff
@@ -476,7 +476,105 @@ public function __construct()
+                        $suggestionStr .= " (did you mean ".$similarities[0]." or ".$similarities[1]."?)";
+                    } elseif (count($similarities) >= 1) {
+                        $suggestionStr .= " (did you mean ".$similarities[0]."?)";
+                    }
```
This block of code can be moved into a helper function `getFlagSuggestionString($key) : string`, which computes and returns $suggestionStr, returning the suggestion or the empty string.

smileytechguy replied:
👍

smileytechguy commented:
@TysonAndre I _think_ I've fixed all the changes you requested.  I added constants `GETOPT_LONG_OPTIONS` and `GETOPT_SHORT_OPTIONS` for this, as well as the private `getFlagSuggestionString` method.

smileytechguy commented:
Oh wow that last commit was pretty broken, my bad.  Fixing now.

smileytechguy commented:
@TysonAndre Okay, _that_ should be good.  Let me know if you have any other suggestions for the code.

TysonAndre commented:
This was merged in https://github.com/phan/phan/commits/5fdc7137fe167e7a6beaf6019e659487befa7d4b with some tests of expected behavior and adjustments for edge cases

I forgot to mention some details in the original ticket (E.g. case insensitive suggestions such as -h for -H, examples of what expected output should look like, etc)
